src/shared/Core/EnvironmentBase.cs

```diff
@@ -104,6 +103,38 @@ public virtual Process CreateProcess(string path, string args, bool useShellExec
 
             return new Process { StartInfo = psi };
         }
+
+        public bool TryLocateExecutable(string program, out string path)
+        {
+            // On UNIX-like systems we would normally use the "which" utility to locate a program,
+            // but since distributions don't always place "which" in a consistent location we cannot
+            // find it! Oh the irony..
+            // We could also try using "env" to then locate "which", but the same problem exists in
+            // that "env" isn't always in a standard location.
+            //
+            // On Windows we should avoid using the equivalent utility "where.exe" because this will
+            // include the current working directory in the search, and we don't want this.
+            //
+            // The upshot of the above means we cannot use either of "which" or "where.exe" and must
+            // instead manually scan the PATH variable looking for the program.
+            // At least both Windows and UNIX use the same name for the $PATH or %PATH% variable!
```

Comment:
Nit: we seem to be using POSIX in some places (the commit message) and UNIX in others (this comment). Is it ok to use them interchangeably in this way?

Reply:
Yeah, I've not been very strict in the past about POSIX vs UNIX. POSIX is a subset of UNIX-like behaviour/implementation.

Perhaps this should probably be "Unix" and use class names like `UnixEnvironment` etc because, especially in this case, POSIX doesn't _strictly_ define the file system path formats beyond [this](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap03.html#tag_03_266).

VMS paths can be pretty different from UNIX paths. I know @dscho has more opinions on UNIX vs POSIX.

In GCM there are various mapped C Standard Library APIs that are under `src/Core/Interop/Posix/Native` because they're in the POSIX standard, but then there are implementation-specific-isms that live in `Interop/Linux/Native` and `Interop/MacOS/Native`.